osu.Game.Rulesets.Osu/Mods/OsuModSlideIn.cs

```diff
@@ -0,0 +1,120 @@
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Rulesets.Osu.Objects.Drawables;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.Mods
+{
+    public class OsuModSlideIn : ModWithVisibilityAdjustment
+    {
+        public override string Name => "Slide In";
+        public override string Acronym => "SI";
+        public override IconUsage? Icon => FontAwesome.Solid.ArrowsAlt;
+        public override ModType Type => ModType.Fun;
+        public override LocalisableString Description => "The circles are coming for you...";
+        public override double ScoreMultiplier => 1;
+        public override Type[] IncompatibleMods => new[] { typeof(OsuModWiggle), typeof(OsuModMagnetised), typeof(OsuModRepel) };
+
+        protected override void ApplyIncreasedVisibilityState(DrawableHitObject hitObject, ArmedState state) => applySlideIn(hitObject, state);
+        protected override void ApplyNormalVisibilityState(DrawableHitObject hitObject, ArmedState state) => applySlideIn(hitObject, state);
+
+        private int counter = -1;
+
+        private readonly List<Vector2> movementVectors = new List<Vector2>(); // A list to store the movement vectors for each hit object
+        private readonly List<Vector2> originalPositions = new List<Vector2>(); // A list to store the original positions of each hit object
+        private readonly List<double> animationDurations = new List<double>(); // A list to store the animation durations for each hit object
+
+        public override void ApplyToBeatmap(IBeatmap beatmap)
+        {
+            movementVectors.Clear();
+            originalPositions.Clear();
+
+            var osuHitObjects = beatmap.HitObjects
+                                       .OfType<OsuHitObject>().ToList()
+                                       .Where(h => h is HitCircle or Slider)
+                                       .ToList();
+
+            // For each pair of consecutive hit objects in the beatmap...
+            for (int i = 0; i < osuHitObjects.Count - 1; i++)
+            {
+                var currentHitObject = osuHitObjects[i];
+                var nextHitObject = osuHitObjects[i + 1];
+
+                Vector2 effectiveStartPosition = currentHitObject.Position;
+
+                if (currentHitObject is Slider currentSlider)
+                {
+                    effectiveStartPosition = currentSlider.TailCircle.Position;
+                }
+
+                Vector2 effectiveEndPosition = nextHitObject.Position;
+
+                if (nextHitObject is Slider nextSlider)
+                {
+                    effectiveEndPosition = nextSlider.HeadCircle.Position;
+                }
+
+                // Calculate the movement vectors and animation durations
+                Vector2 movementVector = effectiveEndPosition - effectiveStartPosition;
+                double timeDiff = nextHitObject.StartTime - currentHitObject.StartTime;
```

Comment:
This should probably be `nextHitObject.EndTime - currentHitObject.StartTime;`

Reply:
Otherwise the sliders will have a disproportionately slower animation time

Reply:
Actually looks like you can just use `.GetEndTime()` in quite a few of these places, no need to handle `Slider` specially